### PR TITLE
Added the checks before parsing the IPv4 layer

### PR DESCRIPTION
**Pull request by gx740**
I have found the incorrect IPv4 packet that forces the `IPReassembly::processPacket` to throw a `bad_alloc` exception in following line of source code:

`newFrag->fragmentData = new uint8_t[fragWrapper->getIPLayerPayloadSize()];`

Such packets are being received from NIC periodically (not too much). In this case the memory leak occurs (refer to PR #266). I have added the checks before parsing IPv4 layer. The main goal of this PR is to fix this issue. But there are other useful changes (look into the commit description).

Output of valgrind:
```
==5667== Thread 3:
==5667== Argument 'size' of function __builtin_vec_new has a fishy (possibly negative) value: -4
==5667==    at 0x4C2AB28: operator new[](unsigned long) (vg_replace_malloc.c:433)
==5667==    by 0x450A87: pcpp::IPReassembly::processPacket(pcpp::Packet*, pcpp::IPReassembly::ReassemblyStatus&, pcpp::ProtocolType, pcpp::OsiModelLayer) (IPReassembly.cpp:416)
==5667==    by 0x4A0C87: core::TrafficDecoder::doIpReassembly(pcpp::Packet*) (traffic_decoder.cpp:61)
==5667==    by 0x4A0E21: core::TrafficDecoder::decodePacket(pcpp::RawPacket*, core::TrafficDecoder::DecodingResult&) (traffic_decoder.cpp:135)
==5667==    by 0x43D936: PcapOffline::WorkerThread::processPacket(pcpp::RawPacket*) (worker_thread.cpp:140)
==5667==    by 0x43DFC6: PcapOffline::WorkerThread::run() (worker_thread.cpp:79)
==5667==    by 0x4A5F2E: execute_native_thread_routine (in /home/utech/test-offline/bin/test-offline)
==5667==    by 0x507FE64: start_thread (in /usr/lib64/libpthread-2.17.so)
==5667==    by 0x64C188C: clone (in /usr/lib64/libc-2.17.so)

```

### CODE REVIEW

File: Packet++/src/NullLoopbackLayer.cpp [hunk 1]
```diff
@@ -25,26 +26,17 @@ NullLoopbackLayer::NullLoopbackLayer(uint32_t family)
...
-    {
-		if ((family & 0x000000FF) == 0 && (family & 0x0000FF00) < 0x00000600)
-		{
-			family = BSWAP16(family & 0xFFFF);
-		}
-    }
+			family = BSWAP32(family);
+	}
+	else if ((family & 0x000000FF) == 0 && (family & 0x0000FF00) < 0x00000600)
+		family = BSWAP16(family & 0xFFFF);
```
Comment (seladb) [inline Packet++/src/NullLoopbackLayer.cpp]: why removing the parentheses? It's more readable that way...

File: Packet++/src/EthLayer.cpp [hunk 2]
```diff
@@ -20,11 +20,14 @@
...
 EthLayer::EthLayer(const MacAddress& sourceMac, const MacAddress& destMac, uint16_t etherType) : Layer()
 {
-	m_DataLen = sizeof(ether_header);
-	m_Data = new uint8_t[m_DataLen];
-	memset(m_Data, 0, m_DataLen);
+	const size_t headerLen = sizeof(ether_header);
+	m_DataLen = headerLen;
+	m_Data = new uint8_t[headerLen];
+	memset(m_Data, 0, headerLen);
+
```
[the remark this answers is not in the document]
Comment (gx740) [reply Packet++/src/EthLayer.cpp]: These changes allow the compiler to produce more effective binary code. Let's get old code:
```
m_DataLen = sizeof(ether_header);
m_Data = new uint8_t[m_DataLen];
memset(m_Data, 0, m_DataLen);
```
Compiler will generally generate the next operations:

1.  assign constant to `m_DataLen`
2.  read value from `m_DataLen` and pass the value as argument of operator `new[]`
3.  read value from `m_DataLen` and pass the value as argument of function `memset`

Depending on how smart the compiler is, some operations can be optimized. The binary code that was generated by g++ 8.3.1 for the old source code:
```
// operation 1
0x0000000000477f07 <+71>:    movq   $0xe,0x10(%rdi)
// operation 2
0x0000000000477f0f <+79>:    mov    $0xe,%edi
0x0000000000477f14 <+84>:    callq  0x405a90 <_Znam@plt>
// operation 3
0x0000000000477f19 <+89>:    mov    %rax,0x8(%rbx)
0x0000000000477f1d <+93>:    mov    0x10(%rbx),%rdx
0x0000000000477f21 <+97>:    xor    %esi,%esi
0x0000000000477f23 <+99>:    mov    %rax,%rdi
0x0000000000477f26 <+102>:   callq  0x405700 <memset@plt>
```
In this case the compiler could not optimize operation 3.

The binary code for the new source code:
```
// operation 1
0x0000000000477f07 <+71>:    movq   $0xe,0x10(%rdi)
// operation 2
0x0000000000477f0f <+79>:    mov    $0xe,%edi
0x0000000000477f14 <+84>:    callq  0x405a90 <_Znam@plt>
// operation 3
0x0000000000477f19 <+89>:    mov    %rax,0x8(%rbx)
0x0000000000477f1d <+93>:    mov    0x0(%r13),%edx
0x0000000000477f21 <+97>:    mov    %ebp,%ecx
```
In this case the call of `memset` was removed.

This file contains the generated binary code for both cases:
[binary.zip](https://github.com/seladb/PcapPlusPlus/files/3787297/binary.zip)

File: Packet++/src/GreLayer.cpp [hunk 3]
```diff
@@ -255,9 +259,10 @@ size_t GreLayer::getHeaderLen()
 
 GREv0Layer::GREv0Layer()
 {
-	m_DataLen = sizeof(gre_basic_header);
-	m_Data = new uint8_t[m_DataLen];
-	memset(m_Data, 0, m_DataLen);
+	const size_t headerLen = sizeof(gre_basic_header);
+	m_DataLen = headerLen;
+	m_Data = new uint8_t[headerLen];
+	memset(m_Data, 0, headerLen);
```
[the remark this answers is not in the document]
Comment (gx740) [reply Packet++/src/GreLayer.cpp]: The same optimizations as for EthLayer::EthLayer().

File: Packet++/src/IPv4Layer.cpp [hunk 4]
```diff
@@ -160,12 +160,14 @@ IPv4Option IPv4OptionBuilder::build() const
...
 /// ~~~~~~~~~
 
+
 void IPv4Layer::initLayer()
 {
-	m_DataLen = sizeof(iphdr);
-	m_Data = new uint8_t[m_DataLen];
+	const size_t headerLen = sizeof(iphdr);
+	m_DataLen = headerLen;
+	m_Data = new uint8_t[headerLen];
```
[the remark this answers is not in the document]
Comment (gx740) [reply Packet++/src/IPv4Layer.cpp]: The same optimizations as for EthLayer::EthLayer().

File: Packet++/src/NullLoopbackLayer.cpp (hunk 1 above)
Comment (gx740) [reply Packet++/src/NullLoopbackLayer.cpp]: The existence of braces increase the number of rows in source code. This results in the need to scroll the screen. I automatically skip the braces for such cases. I recognize that the code formatting depends on personal preferences. From my point of view the second version looks more lightweight. But I can return back the previous version (with braces).
```
uint32_t NullLoopbackLayer::getFamily()
{
  uint32_t family = *(uint32_t*)m_Data;
  if ((family & 0xFFFF0000) != 0)
  {
    if ((family & 0xFF000000) == 0 && (family & 0x00FF0000) < 0x00060000)
    {
      family >>= 16;
    }
    else
    {
      family = BSWAP32(family);
    }
  }
  else if ((family & 0x000000FF) == 0 && (family & 0x0000FF00) < 0x00000600)
  {
    family = BSWAP16(family & 0xFFFF);
  }

  return family;
}
```
```
uint32_t NullLoopbackLayer::getFamily()
{
  uint32_t family = *(uint32_t*)m_Data;
  if ((family & 0xFFFF0000) != 0)
  {
    if ((family & 0xFF000000) == 0 && (family & 0x00FF0000) < 0x00060000)
      family >>= 16;
    else
      family = BSWAP32(family);
  }
  else if ((family & 0x000000FF) == 0 && (family & 0x0000FF00) < 0x00000600)
    family = BSWAP16(family & 0xFFFF);

  return family;
}
```

File: Packet++/src/GreLayer.cpp [hunk 5]
```diff
@@ -206,25 +206,29 @@ void GreLayer::parseNextLayer()
...
 	gre_basic_header* header = (gre_basic_header*)m_Data;
+	uint8_t *payload = m_Data + headerLen;
+	size_t payloadLen = m_DataLen - headerLen;
+
 	switch (ntohs(header->protocol))
 	{
 	case PCPP_ETHERTYPE_IP:
-		m_NextLayer = new IPv4Layer(m_Data + headerLen, m_DataLen - headerLen, this, m_Packet);
+		if(IPv4Layer::isDataValid(payload, payloadLen))
+			m_NextLayer = new IPv4Layer(payload, payloadLen, this, m_Packet);
```
[the remark this answers is not in the document]
Comment (gx740) [reply Packet++/src/GreLayer.cpp]: Done

File: Packet++/src/EthLayer.cpp [hunk 6]
```diff
@@ -38,31 +41,35 @@ void EthLayer::parseNextLayer()
...
 	ether_header* hdr = getEthHeader();
+	uint8_t *payload = m_Data + sizeof(ether_header);
+	size_t payloadLen = m_DataLen - sizeof(ether_header);
+
 	switch (ntohs(hdr->etherType))
 	{
 	case PCPP_ETHERTYPE_IP:
-		m_NextLayer = new IPv4Layer(m_Data + sizeof(ether_header), m_DataLen - sizeof(ether_header), this, m_Packet);
+		if(IPv4Layer::isDataValid(payload, payloadLen))
+			m_NextLayer = new IPv4Layer(payload, payloadLen, this, m_Packet);
```
[the remark this answers is not in the document]
Comment (gx740) [reply Packet++/src/EthLayer.cpp]: Done

File: Packet++/src/GreLayer.cpp [hunk 7]
```diff
@@ -557,16 +564,20 @@ void PPP_PPTPLayer::parseNextLayer()
...
 
+	uint8_t *payload = m_Data + headerLen;
+	size_t payloadLen = m_DataLen - headerLen;
+
 	switch (ntohs(getPPP_PPTPHeader()->protocol))
 	{
 	case PCPP_PPP_IP:
-		m_NextLayer = new IPv4Layer(m_Data + headerLen, m_DataLen - headerLen, this, m_Packet);
+		if(IPv4Layer::isDataValid(payload, payloadLen))
+			m_NextLayer = new IPv4Layer(payload, payloadLen, this, m_Packet);
```
[the remark this answers is not in the document]
Comment (gx740) [reply Packet++/src/GreLayer.cpp]: Done

File: Packet++/src/NullLoopbackLayer.cpp (hunk 1 above)
Comment (gx740) [reply Packet++/src/NullLoopbackLayer.cpp]: The braces have been returned back

File: Packet++/src/EthLayer.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (seladb) [reply Packet++/src/EthLayer.cpp]: got it. I think there are a lot of similar optimizations we can do in the code, but so far I haven't got to that level. But I think we can keep it

File: Packet++/src/NullLoopbackLayer.cpp (hunk 1 above)
Comment (seladb) [reply Packet++/src/NullLoopbackLayer.cpp]: Thanks!

**Pull request merged**